orttraining/orttraining/python/training/ortmodule/ortmodule.py

```diff
@@ -269,3 +281,39 @@ def named_modules(self, *args, **kwargs):
         """Override :meth:`~torch.nn.Module.named_modules`"""
 
         yield from self._torch_module.named_modules(*args, **kwargs)
+
+    def __getattr__(self, name: str):
+        if '_torch_module' in self.__dict__:
+            # If attribute is not found in ORTModule, it must be present in the
+            # user's torch.nn.Module. Forward the call to the user's model.
+            return getattr(self.module, name)
+        else:
+            super(ORTModule, self).__getattr__(name)
+
+    def __setattr__(self, name: str, value) -> None:
+
+        def _setattr_on_user_module(name, value):
+            # Set the attribute on the user's original module
+            setattr(self.module, name, value)
+            # Signal to execution manager to re-export the model.
+            # Re-export will be avoided if _skip_check is enabled.
+            if isinstance(self._torch_module, TorchModuleORT):
+                for training_mode in [False, True]:
+                    self._torch_module._execution_manager(training_mode).mark_for_reexport()
+
+        if isinstance(value, torch.nn.Parameter) or isinstance(value, torch.nn.Module) \
+            or isinstance(value, torch.Tensor):
+            # Any attribute which is either a torch Parameter, Module or
+            # Tensor must be set on the user's model
+            _setattr_on_user_module(name, value)
+        elif name in self.__dict__:
+            # If the name is an attribute of ORTModule, update only ORTModule
+            if '_torch_module' in self.__dict__ and hasattr(self.module, name):
+                warnings.warn(f"User Module attribute name {name} collides with ORTModule attribute name.")
+            self.__dict__[name] = value
+        elif '_torch_module' in self.__dict__:
+            # If the name is an attribute of user model, or is a new attribute, update there.
+            _setattr_on_user_module(name, value)
+        else:
+            # Setting any new attributes should be done on ORTModule only when 'torch_module' is not defined
+            self.__dict__[name] = value
```

Comment:
when that happens and why we need it?

Correct me if I am wrong, but there should be two cases: 1) Existing ORTModule methods/attributes can be updated in runtime and 2) everything else should go direct to the use model. This block has 4 conditions, but if feels it could be 2, only

main if from line 309 to update ORTModule and an else changing user model.

Reply:
```__setattr__``` gets called whenever an attempt is made to set any attribute on the model. There are two scenarios when `ORTModule` needs to be updated.
- When the attribute is already set on ```ORTModule``` by a previous ```__setattr__``` call. Examples include calling ```model.train()``` that should update the ```ORTModule``` ```training``` attribute. This is what the second elif block tackles.
- When new attributes are being set on ```ORTModule```. This happens when we call super constructor in ```__init__``` and it sets the attributes like ```training, _modules, _parameters, _buffers, _backward_hooks...``` for the first time. This is what the final else block tackles.

That is why we need both these blocks.